src/internal_modules/roc_core/mpsc_queue_impl.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/mpsc_queue_impl.h
+//! @brief Multi-producer single-consumer queue internal implementation.
+
+#ifndef ROC_CORE_MPSC_QUEUE_IMPL_H_
+#define ROC_CORE_MPSC_QUEUE_IMPL_H_
+
+#include "roc_core/mpsc_queue_node.h"
+
+namespace roc {
+namespace core {
+
+//! Multi-producer single-consumer queue internal implementation class.
+//!
+//! Provides only push/pop functionality. Ownership is left up to the main MpscQueue
+//! class.
+class MpscQueueImpl {
+public:
+    MpscQueueImpl();
+
+    ~MpscQueueImpl();
+
+    //! Add object to the end of the queue.
+    void push_back(MpscQueueNode::MpscQueueData* node);
+
+    //! Remove object from the beginning of the queue.
+    template <bool CanSpin> MpscQueueNode::MpscQueueData* pop_front() {
+        return pop_node_(CanSpin);
+    }
+
+private:
+    typedef MpscQueueNode::MpscQueueData MpscQueueData;
+
+    void push_node_(MpscQueueData* node);
+
+    MpscQueueData* pop_node_(bool can_spin);
```

Comment:
Unfortunately, when change_owner_() is moved to this class and these functions, they result in problems, because we are pushing stub_ at some point while popping.

Reply:
Let me know if I should change the behaviour a bit more to avoid this extra indirection.

Reply:
Now after moving change_owner_ this indirection makes sense again, looks fine.